### PR TITLE
Ubisoft changes

### PR DESCRIPTION
**Pull request by dave-juicelabs**
- Juicify has new options --queue-timeout, --on-queue-timeout, --on-connection-error supporting the Ubisoft use case of continuing even when Juice is not available
- Errors are cleaned up a bit in juicify
- juiceclient.dll has GetVersion so we can access Juice version from the Juice-Labs code

### CODE REVIEW

File: cmd/juicify/app/juicify.go [hunk 1]
```diff
@@ -5,66 +5,151 @@ package app
...
-	DisableCompression bool `json:"disableCompression,omitempty"`
+				case <-timeout.C:
+					timeoutChannel <- struct{}{}
+					return nil
+				}
+			})
+		}
 
-	Headless bool `json:"headless,omitempty"`
+		ticker := time.NewTicker(5 * time.Second)
```
Comment (ykagan) [inline cmd/juicify/app/juicify.go:103]: should we poll more frequently than this?

File: pkg/restapi/types.go [hunk 2]
```diff
@@ -125,10 +125,10 @@ type SessionUpdate struct {
 }
 
 type AgentUpdate struct {
-	Id             string                   `json:"id"`
-	State          string                   `json:"state"`
-	SessionsUpdate map[string]SessionUpdate `json:"sessions"`
-	Gpus           []GpuMetrics             `json:"gpus"`
+	Id      string                   `json:"id"`
+	State   string                   `json:"state"`
+	Updates map[string]SessionUpdate `json:"sessions"`
```
Comment (ykagan) [inline pkg/restapi/types.go:130]: nit: its a bit confusing to have this be update but the the json string to be 'sessions'?

File: cmd/juicify/app/juicify.go [hunk 3]
```diff
@@ -5,66 +5,151 @@ package app
...
+			case <-ticker.C:
+				session, err = api.GetSessionWithContext(group.Ctx(), id)
+			}
 
-	SwapChainBuffers int `json:"swapChainBuffers,omitempty"`
+			if err != nil {
+				logger.Info("Cancelling session")
 
-	WaitForDebugger bool `json:"waitForDebugger,omitempty"`
+				api.CancelSession(id)
```
Comment (ykagan) [inline cmd/juicify/app/juicify.go:121]: can we also cancel the session on exit?

**Review by ykagan — APPROVED** [review]

File: cmd/agent/app/controller.go [hunk 4]
```diff
@@ -46,133 +46,136 @@ type controllerData struct {
...
-					})
-
-				case <-ticker.C:
-					// Update our state from what is on the controller
-					controllerAgent, err := agent.api.GetAgentWithContext(group.Ctx(), agent.Id)
-					if err != nil {
-						return err
-					}
+	// When connected to the controller, the agent must not allow requests
+	agent.Server.SetCreateEndpoint(RequestSessionName, nil)
```
[the remark this answers is not in the document]
Comment (dave-juicelabs) [reply cmd/agent/app/controller.go:88]: Done

File: cmd/juicify/app/juicify.go (hunk 1 above)
Comment (dave-juicelabs) [reply cmd/juicify/app/juicify.go:103]: All the explicit values should be configurable in some way. This was just a reasonable value.

File: cmd/juicify/app/juicify.go (hunk 3 above)
Comment (dave-juicelabs) [reply cmd/juicify/app/juicify.go:121]: The only concern with that is without the Windows Jobs, in applications like 3DMark, juicify will start a launcher which launches something else and then exits which then triggers juicify to exit. We fix that and yes we can cancel the session on juicifys exit.

File: pkg/restapi/types.go (hunk 2 above)
Comment (dave-juicelabs) [reply pkg/restapi/types.go:130]: Yes, the name SessionsUpdate seems semantically incorrect and it was bugging me so I switched it to Updates. Overall, I am not happy with this update structure at all anymore. I think a more generic format would be good but I am not sure what that would look like yet.

**Review by canadacow — APPROVED** [review]
LGTM

**Review by canadacow — APPROVED** [review]
@ykagan , I trust your review on this. Otherwise LGTM.

**Pull request merged**